### PR TITLE
Tylake/configuration

### PR DESCRIPTION
**Pull request by tylorhl**
Apologies for the large PR.

This PR serves a couple of primary purposes:
* Decouple ServiceFabric configuration requirement to allow more infrastructure independence
  * This is done by adding a provider that adds to IConfiguration
* Add Auth* capabilities to be used directly on and in each of the service controllers

Some additions to this:
* Integrate OneBox scenario without separate codebases
* Changed `ILogger` signatures to use the `ILogger<>` interface and changed the MEF DI to accommodate it

General notes:
* No real changes to business logic; mostly entrypoint modification for settings
* Use of new auth with Attributes is optional to allow gradual moving of services
  * Set up to take into account backwards-compatibility with current ServiceFabric setup scenario so that the same code should work as-is in ServiceFabric behind the Gateway

### CODE REVIEW

File: Services/DataX.Flow/Flow.ManagementService/appsettings.Development.json [hunk 1]
```diff
@@ -5,5 +5,36 @@
...
+      "CosmosDBConfigDatabaseName": "",
+      "MefBlobDirectory": null,
+      "MefContainerName": null,
+      "MefStorageAccountName": null,
+      "ServiceKeyVaultName": "",
+
+      // EnableOneBox == true settings
+      "LocalRoot": "",
+      "MetricsHttpEndpoint": "http://localhost:2020/",
+      "SparkHome": ""
```
Comment (dataxpd) [inline Services/DataX.Flow/Flow.ManagementService/appsettings.Development.json]: Need to add EnableOneBox

File: Services/DataX.Flow/Flow.ManagementService/appsettings.json [hunk 2]
```diff
@@ -4,5 +4,22 @@
...
+    "ServiceEnvironment": {
+      "EnableOneBox": false,
+      "AppInsightsIntrumentationKey": "",
+      "CosmosDBConfigCollectionName": "",
+      "CosmosDBConfigConnectionString": "",
+      "CosmosDBConfigDatabaseName": "",
+      "MefBlobDirectory": null,
+      "MefContainerName": null,
+      "MefStorageAccountName": null,
+      "ServiceKeyVaultName": ""
```
Comment (dataxpd) [inline Services/DataX.Flow/Flow.ManagementService/appsettings.json]: Need to add EnableOneBox

File: Services/DataX.Config/DataX.Config/Utility/LoggerExportDescriptorProvider.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        private readonly bool _hasInstances;
+
+        public LoggerExportDescriptorProvider(object[] instances, ILoggerFactory loggerFactory)
+            : base(instances)
+        {
+            _hasInstances = instances?.Length > 0;
+            _loggerFactory = loggerFactory;
+        }
+
+        public override IEnumerable<ExportDescriptorPromise> GetExportDescriptors(CompositionContract contract, DependencyAccessor descriptorAccessor)
```
Comment (dataxpd) [inline Services/DataX.Config/DataX.Config/Utility/LoggerExportDescriptorProvider.cs]: Please add some comment headers

**Comment by tylorhl** [comment]
> Another file which requires to be updated with EnableOneBox variable: https://github.com/microsoft/data-accelerator/blob/master/Services/DataX.Flow/Flow.ManagementService/PackageRoot/ServiceManifest.xml-Add:
> 
> ```
> <Parameter Name="EnableOneBox" Value="" MustOverride="true" />
> ```

It doesn't appear that it needs it. Did you mean [Settings.xml](https://github.com/microsoft/data-accelerator/blob/master/Services/DataX.Flow/Flow.ManagementService/PackageRoot/Config/Settings.xml)?

File: Services/DataX.Flow/Flow.ManagementService/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -23,9 +25,9 @@ public static class ServiceCollectionExtensions
         /// <param name="exportTypes">Types being exported</param>
         /// <param name="instanceExports">Instances that are created outside that need to be added to the container</param>
         /// <returns></returns>
-        public static IServiceCollection AddMefExportsFromAssemblies(this IServiceCollection services, ServiceLifetime lifetime, IEnumerable<Assembly> assemblies, Type[] exportTypes, object[] instanceExports)
+        public static IServiceCollection AddMefExportsFromAssemblies(this IServiceCollection services, ServiceLifetime lifetime, IEnumerable<Assembly> assemblies, Type[] exportTypes, object[] instanceExports, ILoggerFactory loggerFactory = null, bool local = false)
         {          
-            var configuration = new ContainerConfiguration().WithAssemblies(assemblies).WithProvider(new InstanceExportDescriptorProvider(instanceExports));
+            var configuration = new ContainerConfiguration().WithAssemblies(assemblies).WithProvider(new LoggerExportDescriptorProvider(instanceExports, loggerFactory));
```
Comment (vijayupadya) [inline Services/DataX.Flow/Flow.ManagementService/ServiceCollectionExtensions.cs]: Should probably rename this class. Name gives the impression that this handles only logger instances which is what I assumed initially, but then after looking at the implementation saw that this handles non-logger instances as-well.

File: Services/DataX.Flow/Flow.ManagementService/PackageRoot/Config/Settings.xml [hunk 5]
```diff
@@ -14,5 +14,6 @@ Licensed under the MIT License
     <Parameter Name="MefStorageAccountName" Value="" />
     <Parameter Name="MefContainerName" Value="" />
     <Parameter Name="MefBlobDirectory" Value="" />
+    <Parameter Name="EnableOneBox" Value="" />
```
Comment (s-tuli) [inline Services/DataX.Flow/Flow.ManagementService/PackageRoot/Config/Settings.xml]: Can you also add: Value="" **MustOverride**="true" />

**Pull request merged**